File: js/src/subprotocol.ts
export interface SubprotocolMessage {
  subprotocolId: number;
  sourceClientId: string;
  /** Owned complete message, copied across the WASM boundary. */
  bytes: Uint8Array;
}

export interface SubprotocolDevice {
  /** Requires a provider-capable native companion; hosted proxies reject it.
   * Register mirrored peer/state listeners before the first RPC sync. Adding
   * them later can reconnect the browser RPC and invalidate subscriptions. */
  enableSubprotocol(id: number, listener: (message: SubprotocolMessage) => void | Promise<void>): Promise<SubprotocolSubscription>;
}

/** Internal WASM bridge. */
export interface SubprotocolBridge {
  subprotocolOperation(operation: string, handle: number, argument: unknown): Promise<any>;
}

/** A registration on one RPC session. Reopen explicitly after disconnection.
 * Up to 16 registrations/session, 65535 bytes/frame, and 64 queued messages or
 * 1 MiB/registration. Overflow rejects `closed`; frames are never coalesced. */
export class SubprotocolSubscription {
  readonly subprotocolId: number;
  /** Resolves after unsubscribe; rejects on transport, queue, or listener failure. */
  readonly closed: Promise<void>;
  private active = true;
  private release?: Promise<void>;
  private bridge: SubprotocolBridge;
  private handle: number;
  constructor(bridge: SubprotocolBridge, handle: number, id: number, listener: (message: SubprotocolMessage) => void | Promise<void>) {
    this.bridge = bridge;
    this.handle = handle;
    this.subprotocolId = id;
    this.closed = this.receive(listener);
    // Applications may inspect .closed later without an unhandled rejection.
    void this.closed.catch(() => {});
  }
  private async receive(listener: (message: SubprotocolMessage) => void | Promise<void>): Promise<void> {
    try {
      while (this.active) {
        const message = await this.bridge.subprotocolOperation("receive", this.handle, null);
        if (!this.active) break;
        if (!(message.bytes instanceof Uint8Array)) throw new TypeError("Invalid subprotocol frame from WASM");
        await listener({subprotocolId: this.subprotocolId, sourceClientId: message.sourceClientId, bytes: message.bytes.slice()});
      }
    } catch (error) {
      if (this.active) throw error;
    } finally {
      await this.close();
    }
  }
  private requireActive(): void {
    if (!this.active) throw new Error("Subprotocol subscription is closed");
  }
  /** true means enqueued. Use an application ACK to establish receipt. */
  async send(destinationClientId: string, bytes: Uint8Array): Promise<boolean> {
    this.requireActive();
    if (!(bytes instanceof Uint8Array)) throw new TypeError("Expected Uint8Array");
    if (bytes.length > 65535) throw new RangeError("Subprotocol frame exceeds 65535 bytes");
    return this.bridge.subprotocolOperation("send", this.handle, {destinationClientId, bytes: bytes.slice()});
  }
  /** null means unanswered/unavailable; [] means the peer advertised no protocols. */
  async querySubprotocols(destinationClientId: string, timeoutMillis = 10000): Promise<number[] | null> {
    this.requireActive();
    if (!Number.isSafeInteger(timeoutMillis) || timeoutMillis < 1 || timeoutMillis > 60000) throw new RangeError("timeoutMillis must be between 1 and 60000");
    return this.bridge.subprotocolOperation("query", this.handle, {destinationClientId, timeoutMillis});
  }
  /** Stops delivery immediately and releases the native registration once. */
  close(): Promise<void> {
    this.active = false;
    this.release ??= this.bridge.subprotocolOperation("release", this.handle, null);
    return this.release;
  }
}

export function attachSubprotocolAPI<T extends object>(device: T): T & SubprotocolDevice {
  const bridge = device as unknown as SubprotocolBridge;
  return Object.assign(device, {
    async enableSubprotocol(id: number, listener: (message: SubprotocolMessage) => void | Promise<void>): Promise<SubprotocolSubscription> {
      if (!Number.isInteger(id) || id < 1024 || id > 65535) throw new RangeError("Application subprotocol id must be between 1024 and 65535");
      if (typeof listener !== "function") throw new TypeError("Expected subprotocol listener");
      if (typeof bridge.subprotocolOperation !== "function") throw new Error("Loaded WASM lacks subprotocol RPC support; rebuild the SDK");
      const handle = await bridge.subprotocolOperation("open", 0, id);
      return new SubprotocolSubscription(bridge, handle.id, id, listener);
    },
  });
}

File: js/subprotocol.go
//go:build js

package main

import (
	"context"
	"errors"
	"net"
	"syscall/js"

	"github.com/urnetwork/sdk/v2026"
)

type jsSubprotocol interface {
	Receive(context.Context) (*sdk.Id, []byte, error)
	Send(context.Context, *sdk.Id, []byte) (bool, error)
	Query(context.Context, *sdk.Id, int64) ([]int32, bool, error)
	Close() error
}

type jsOpenSubprotocol func(context.Context, int32) (jsSubprotocol, error)

// One dispatcher and the existing bounded handle registry per device. Copies
// happen synchronously on entry/exit; Go never retains a caller's typed array.
func jsBindSubprotocolDevice(parent context.Context, open jsOpenSubprotocol, m map[string]any) *jsSocketHandles {
	ctx, cancel := context.WithCancel(parent)
	h := &jsSocketHandles{values: make(map[int]any), done: make(map[int]chan struct{}), ctx: ctx, cancel: cancel}
	go func() { <-ctx.Done(); h.close() }()
	m["subprotocolOperation"] = js.FuncOf(func(_ js.Value, args []js.Value) any {
		if len(args) != 3 || args[0].Type() != js.TypeString || args[1].Type() != js.TypeNumber {
			return jsRejected(errors.New("subprotocolOperation requires operation, handle, and argument"))
		}
		op, handle, arg := args[0].String(), args[1].Int(), args[2]
		var protocol int32
		var timeout int64
		var destination *sdk.Id
		var data []byte
		if op == "open" {
			if arg.Type() != js.TypeNumber || arg.Float() != float64(arg.Int()) || arg.Int() < int(sdk.SubprotocolReservedLimit) || arg.Int() > 65535 {
				return jsRejected(errors.New("invalid application subprotocol id"))
			}
			protocol = int32(arg.Int())
		}
		if op == "send" || op == "query" {
			if arg.Type() != js.TypeObject || arg.IsNull() || arg.Get("destinationClientId").Type() != js.TypeString {
				return jsRejected(errors.New("destinationClientId is required"))
			}
			var err error
			destination, err = sdk.ParseId(arg.Get("destinationClientId").String())
			if err != nil {
				return jsRejected(err)
			}
			if op == "send" {
				value := arg.Get("bytes")
				if !value.InstanceOf(js.Global().Get("Uint8Array")) || value.Length() > 65535 {
					return jsRejected(errors.New("expected Uint8Array of at most 65535 bytes"))
				}
				data = make([]byte, value.Length())
				js.CopyBytesToGo(data, value)
			} else {
				value := arg.Get("timeoutMillis")
				if value.Type() != js.TypeNumber || value.Float() != float64(value.Int()) || value.Int() < 1 || value.Int() > 60000 {
					return jsRejected(errors.New("timeoutMillis must be between 1 and 60000"))
				}
				timeout = int64(value.Int())
			}
		}
		return jsPromise(func(resolve func(any), reject func(error)) {
			var result any
			var err error
			if op == "open" {
				var channel jsSubprotocol
				channel, err = open(ctx, protocol)
				if err == nil {
					result, err = h.add(channel)
				}
			} else {
				h.mu.Lock()
				resource := h.values[handle]
				if op == "release" {
					delete(h.values, handle)
					if done := h.done[handle]; done != nil {
						close(done)
						delete(h.done, handle)
					}
				}
				h.mu.Unlock()
				channel, _ := resource.(jsSubprotocol)
				if op == "release" {
					if channel != nil {
						err = channel.Close()
					}
				} else if channel == nil {
					err = net.ErrClosed
				} else {
					switch op {
					case "receive":
						var source *sdk.Id
						var frame []byte
						source, frame, err = channel.Receive(ctx)
						if err == nil {
							bytes := js.Global().Get("Uint8Array").New(len(frame))
							js.CopyBytesToJS(bytes, frame)
							result = map[string]any{"sourceClientId": source.String(), "bytes": bytes}
						}
					case "send":
						result, err = channel.Send(ctx, destination, data)
					case "query":
						var ids []int32
						var ok bool
						ids, ok, err = channel.Query(ctx, destination, timeout)
						if err == nil && ok {
							out := make([]any, len(ids))
							for i, id := range ids {
								out[i] = id
							}
							result = out
						}
					default:
						err = errors.New("unknown subprotocol operation")
					}
				}
			}
			if err != nil {
				reject(err)
			} else {
				resolve(result)
			}
		})
	})
	return h
}
